src/plugins/gs-plugin-external-apps.c

```diff
@@ -0,0 +1,712 @@
+	if (!gs_plugin_run_command (repo_dir, argv2, error))
+		return FALSE;
+
+	return TRUE;
+}
+
+static gboolean
+flatpak_repo_build_export (const char *repo_dir,
+			   const char *repo_name,
+			   GError **error)
+{
+	const char *argv[] = {"flatpak", "build-export",  "--runtime",
+			      repo_name, repo_dir, NULL};
+	return gs_plugin_run_command (NULL, argv, error);
+}
+
+static gboolean
+flatpak_remote_add (const char *repo_dir,
+		    const char *repo_name,
+		    GError **error)
+{
+	const char *argv[] = {"flatpak", "remote-add",  "--user",
+			      "--no-gpg-verify", repo_name, repo_dir, NULL};
+	return gs_plugin_run_command (NULL, argv, error);
+}
+
+static gboolean
+flatpak_remote_delete (const char *repo_name,
+		       GError **error)
+{
+	const char *argv[] = {"flatpak", "remote-delete",  "--user", "--force",
+			      repo_name, NULL};
+	return gs_plugin_run_command (NULL, argv, error);
+}
+
+static gboolean
+gs_plugin_add_runtime_deb_asset (const gchar *build_dir,
+				 const gchar *repo_dir,
+				 const char *asset_path,
+				 GError **error)
+{
+	const char *argv[] = {"ar", "x",  asset_path, NULL};
+	const char *argv2[] = {"tar", "xzf",  "data.tar.gz", "-C", repo_dir,
+			       NULL};
+
+	if (!gs_plugin_run_command (build_dir, argv, error)) {
+		return FALSE;
+	}
+
+	if (!gs_plugin_run_command (build_dir, argv2, error)) {
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+gs_plugin_add_runtime_asset (GsPlugin *plugin,
+			     GsApp *app,
+			     const gchar *build_dir,
+			     const gchar *repo_dir,
+			     const char *runtime,
+			     const char *asset,
+			     GError **error)
+{
+	g_autofree gchar *content_type = NULL;
+	g_autoptr(GFile) download_file = NULL;
+
+	char *download_name = gs_plugin_download_asset (plugin, app, runtime,
+							asset, error);
+
+	if (!download_name) {
+		return FALSE;
+	}
+
+	download_file = g_file_new_for_path (download_name);
+	content_type = gs_utils_get_content_type (download_file, NULL,
+						  error);
+
+	if (!content_type) {
+		return FALSE;
+	}
+
+	g_debug ("Adding runtime asset with type '%s'", content_type);
+
+	if (g_strcmp0 (content_type, "application/x-deb") == 0) {
+		gs_plugin_add_runtime_deb_asset (build_dir, repo_dir,
+						 download_name, error);
+	} else {
+		g_debug ("Cannot deal with asset type '%s'", content_type);
+
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+gs_plugin_build_runtime (GsPlugin *plugin,
+			 GsApp *app,
+			 GsApp *runtime,
+			 GError **error)
+{
+	gboolean ret;
+	g_autofree char *tmp_dir;
+	const char *template = "3rdparty_XXXXXX";
+	const char *build_dir = "3rd_party_repo";
+	g_autofree char *repo_dir = NULL;
+	g_autofree char *tmp_build_dir = NULL;
+	g_autofree char *repo_name = NULL;
+	g_autofree char *dir_basename = NULL;
+	const char *runtime_name = gs_app_get_id_no_prefix (runtime);
+	const char *runtime_url = gs_app_get_metadata_item (runtime,
+							    METADATA_URL);
+
+	tmp_dir = g_dir_make_tmp (template, error);
```

Comment:
Unfortunately on our system, `/tmp` is in RAM, so this will take a lot of precious memory.
You should use a location in `/var` such as `/var/tmp` or perhaps even better somewhere in the user's home